### DISCUSSION TITLE
GET, POST等参数提取

### DISCUSSION THREAD

**Discussion by weq95** [tokio-rs/axum#1860] [Q&A] [answered]
pub struct AppState {
    redis: std::sync::Arc<RedisConn>,
    mysql: std::sync::Arc<MySqlConn>,
}

pub struct User {
    pub id: i64,
    pub name: String,
}

/// HTTP请求参数提取示例
/// GET 请求: https://example.com/orders/user/:id?page=1&page_per=15
pub fn example(
    axum::extract::Path(user_id): axum::extract::Path<i64>,
    axum::extract::Query(pagination): axum::extract::Query<std::collections::HashMap<String, usize>>,
    http_header: axum::http::HeaderMap,
    axum::extract::State(state): axum::extract::State<AppState>,
    request: axum::http::Request<axum::body::Body>,
) -> impl IntoResponse {
    // 希望这个示例可以帮助到像我这样的小白
    // 具体参数提取看文档描述: https://docs.rs/axum/0.6.11/axum/extract/index.html
    let user = request.extensions().get::<User>();
    let host = http_header.get("host");
    todo!()
}

**Comment by davidpdrsn** [comment]
What..? I don’t speak Chinese sorry

**Comment by foxsin10** [comment, marked as answer]
He just wrote a request parameter extract example, nothing special.

```Rust
pub struct AppState<T, V> {
    redis: Arc<T>,
    mysql: Arc<V>,
}

pub struct User {
    pub id: i64,
    pub name: String,
}

/// HTTP request parameter extract example
/// GET request: https://example.com/orders/user/:id?page=1&page_per=15
pub fn example<S, B>(
    Path(user_id): Path<String>,
    Query(pagination): Query<HashMap<String, usize>>,
    http_header: HeaderMap,
    State(state): State<S>,
    request: Request<B>,
) -> impl IntoResponse {
    // wish this example can help rookies like me
    // see more details: https://docs.rs/axum/0.6.11/axum/extract/index.html
    let user = request.extensions().get::<String>();
    let host = http_header.get("host");
    todo!()
}
```

**Reply by davidpdrsn** [reply]
You can extract cookies with https://docs.rs/axum-extra/latest/axum_extra/extract/cookie/index.html and the host header with https://docs.rs/axum/latest/axum/extract/struct.Host.html

**Reply by foxsin10** [reply]
Yeah, i saw the doc, great work by the way.
Just translate text and format code from @weq95.
I think we can mark this as answered, as the origin issue seems to provide some sample code about extract.

**Reply by davidpdrsn** [reply]
Alright. Thanks for the help! @weq95 I hope it makes sense to you.
